### DISCUSSION TITLE
Expected meaning of `<Alias>: typing.TypeAlias = typing.ClassVar[<type>]`, with inheritance?

### DISCUSSION THREAD

**Discussion by bzoracler** [python/typing#1172] [Q&A] [answered]
For the following code snippet,

```python
from typing import TypeAlias, ClassVar, get_type_hints

ClassAlias: TypeAlias = ClassVar[str]

class A:
    i: ClassAlias = "g"     # No errors

class B:
    i: ClassVar[str] = "g"  # No errors

class A2(A):
    i: ClassVar[str] = "g"  # <<<<<<< pylance and mypy both flag this as an error

class B2(B):
    i: ClassVar[str] = "g"  # No errors
```

`mypy` and `pylance` are both flagging `A2.i` as a class variable overriding a parent class's *instance* variable (`A.i`):
 - `Class variable "i" overrides instance variable of same name in class "A" Pylance[reportIncompatibleVariableOverride]`
 - `Mypy: Cannot override instance variable (previously declared on base class "A") with class variable`

However, `typing.get_type_hints` show that the type hints on `A` and `B` are identical:

```python
>>> print("type hints of A:", get_type_hints(A))
type hints of A: {'i': typing.ClassVar[str]}
>>> print("type hints of B:", get_type_hints(B))
type hints of B: {'i': typing.ClassVar[str]}
```

I didn't pick up anything about a restricted scope of `typing.ClassVar` in [PEP 526](https://peps.python.org/pep-0526/#class-and-instance-variable-annotations) nor anything about the interaction of `typing.TypeAlias` with `typing.ClassVar` in [PEP 613](https://peps.python.org/pep-0613/), so assumed that we can type alias a `ClassVar[...]` at the module level and use it as a type annotation in the class. What's going on here?

**Comment by erictraut** [comment, marked as answer]
`ClassVar` should not be used outside of a variable declaration within a class. `ClassVar` is not an annotation that describes the _type_ of a variable, but it describes a (non-type) attribute of a variable. Most other languages offer keywords for such attributes, but Python overloads the type annotation mechanism to handle these cases. Other examples include `Final`, `InitVar`, `Required`, and `NotRequired`. These special forms should not be used within type aliases because they are not type annotations. 

The use of `ClassVar` should be flagged as an error when used in a type alias like this. When I implemented support for ClassVar in pyright, it never occurred to me that someone would attempt to use it in a type alias, so there is no check for this error condition. But you should assume that it will not work when used in this manner.

**Reply by erictraut** [reply]
I've filed an [enhancement request in pyright](https://github.com/microsoft/pyright/issues/3412) to properly flag this usage as an error.

**Reply by bzoracler** [reply]
Thanks for the explanation about special forms - it means that this is the same reason why mypy and pylance are giving the same override errors for `<...>:  Alias = ClassVar[<...>]` within a class as well.

```python
from typing import TypeAlias, ClassVar

class A:
    ClassAlias: TypeAlias = ClassVar[int]  # No flagged errors
    i: ClassAlias = 2                      # No flagged errors

class B(A):
    i: ClassVar[int] = 3  # Class variable "i" overrides instance variable of same name in class "A"
```

**Reply by erictraut** [reply]
Yes, the next release of pyright and pylance will properly flag the use of `ClassVar` in a `TypeAlias` declaration as an error. You might want to file a similar enhancement request for mypy so it detects and reports this error condition.
